tests/tools.vitruv.dsls.reactions.tests/src/tools/vitruv/dsls/reactions/formatting2/ReactionLanguageFormatterTest.xtend

```diff
@@ -0,0 +1,207 @@
+				{
+					{
+						newNonRoot.value = "new"
+					}
+				}
+				insertNonRoot(newNonRoot)
+			}
+		}'''
+		assertThat(expected.destroyFormatting().format(), hasEachLineEqualTo(expected))
+	}
+
+	private def String format(String reactionsCodeToFormat) {
+		return reactionsCodeToFormat.parse.serialize(SaveOptions::newBuilder.format().getOptions())
+	}
+
+	private def String destroyFormatting(String textToUnformat) {
+		return textToUnformat.preserveEmptyLines [
+			applyToEachLineRemovingLinebreaks [
+				if (it.isCommentLine) {
+					it.preserveCommentLine()
+				} else {
+					it.mixUpSpacingTerminals()
+				}
+			]
+		]
+	}
+
+	private def String preserveEmptyLines(String text, (String)=>String applicator) {
+		text.replace(lineSeparator + lineSeparator, lineSeparator + lineSeparator, applicator)
+	}
+
+	private def String applyToEachLineRemovingLinebreaks(String text, (String)=>String applicator) {
+		text.split(lineSeparator).map[applicator.apply(it)].join()
+	}
+
+	private def boolean isCommentLine(String line) {
+		return line.isFirstCommentLine || line.isInnerCommentLine
+	}
+
+	private def boolean isFirstCommentLine(extension String line) {
+		return contains("//") || contains("/*")
+	}
+
+	private def boolean isInnerCommentLine(extension String line) {
+		return contains(" *")
+	}
+
+	private def String preserveCommentLine(extension String line) {
+		if (line.isFirstCommentLine) {
+			return lineSeparator + line + lineSeparator
+		} else if (line.isInnerCommentLine) {
+			return line + lineSeparator
+		} else {
+			throw new IllegalArgumentException("line is no comment: " + line)
+		}
+	}
+
+	private def String mixUpSpacingTerminals(String text) {
+		val result = text.replace(" ", lineSeparator + lineSeparator) [
+			replace("::", "  ::  ") [
+				replace('\t', '  ')
+			]
+		]
+		if (result != "}") {
+			return result + " "
+		} else {
+			return result
+		}
+	}
+
+	private def String replace(String text, String original, String replacement,
+		(String)=>String applyOnIntermediatePlaceholderString) {
+		return text.modifyTextUsingPlaceholderChar [inputText, placeholder|
+			val placeholderString = "" + placeholder
+			val textWithPlaceholder = inputText.replace(original, placeholderString)
+			val modifiedTextWithPlaceholder = applyOnIntermediatePlaceholderString.apply(textWithPlaceholder)
+			val textWithReplacement = modifiedTextWithPlaceholder.replace(placeholderString, replacement)
+			return textWithReplacement	
+		]
+	}
+
+	var char nextPlaceholderChar
+
+	@BeforeEach
+	def void restorePlaceholderChars() {
+		nextPlaceholderChar = 14 as char
+	}
+
+	private def String modifyTextUsingPlaceholderChar(String text, (String, char)=>String modifier) {
+		val result = modifier.apply(text, retrievePlaceholderChar())
+		returnPlaceholderChar()
+		return result
+	}
+
+	private def char retrievePlaceholderChar() {
+		return nextPlaceholderChar++
+	}
+
+	private def void returnPlaceholderChar() {
+		nextPlaceholderChar--
+	}
```

Comment:
Maybe one line comment regarding the ASCII char magic here? Like which ASCII characters are expected to be covered.

Reply:
Oh, yes. I have added a comment and also ensured that we do not conflict with actually used ASCII symbols by having a maximum number of placeholder symbols in d4169de.